diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6886a6b4..63b2f30c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -6,7 +6,7 @@ project(Odin2 VERSION 2.4.1 LANGUAGES C CXX ASM)
 # ================== CMake Subdirs ======================
 add_subdirectory(libs/clap-juce-extensions EXCLUDE_FROM_ALL)
 if(UNIX AND NOT APPLE)
-  add_subdirectory(libs/JUCELV2 EXCLUDE_FROM_ALL)
+  add_subdirectory(${JUCE_DIR} "JUCE" EXCLUDE_FROM_ALL)
 else()
   add_subdirectory(libs/JUCE EXCLUDE_FROM_ALL)
 endif()
@@ -36,8 +36,8 @@ add_compile_definitions(
 
 # ==================== Add Plugin =======================
 # Build LV2 only on Linux
-set(JUCE_FORMATS AU VST3 Standalone)
-if(UNIX AND NOT APPLE)
+set(JUCE_FORMATS AU Standalone)
+if(false) #UNIX AND NOT APPLE) # We don't build LV2 on Android
     list(APPEND JUCE_FORMATS LV2)
 endif()
 
@@ -69,7 +69,6 @@ target_compile_definitions(Odin2
     JUCE_USE_CURL=0
     JUCE_VST3_CAN_REPLACE_VST2=0
     JUCE_WEB_BROWSER=0
-    JUCE_MODAL_LOOPS_PERMITTED=1
     JUCE_DISPLAY_SPLASH_SCREEN=0
 )
 
@@ -254,3 +253,52 @@ endforeach()
 if(TARGET ${PROJECT_NAME}_CLAP)
   install(TARGETS ${PROJECT_NAME}_CLAP DESTINATION ${CMAKE_INSTALL_LIBDIR}/clap)
 endif()
+
+# begin JUCEAAP specific -->
+# It is not beautiful, but there does not seem to be any way to add custom include directory for modules.
+include_directories(
+        "${AAP_DIR}/include"
+)
+
+juce_add_module(${AAP_JUCE_DIR}/aap-modules/aap_audio_processors)
+
+if (ANDROID)
+
+    # libcpufeatures
+    add_library("cpufeatures" STATIC "${ANDROID_NDK}/sources/android/cpufeatures/cpu-features.c")
+    set_source_files_properties("${ANDROID_NDK}/sources/android/cpufeatures/cpu-features.c" PROPERTIES COMPILE_FLAGS "-Wno-sign-conversion -Wno-gnu-statement-expression")
+
+    enable_language(ASM)
+
+    # dependencies
+    find_library(log "log")
+    find_library(android "android")
+    find_library(glesv2 "GLESv2")
+    find_library(egl "EGL")
+    set(cpufeatures_lib "cpufeatures")
+    set(oboe_lib "oboe")
+    set(CMAKE_CXX_FLAGS_RELEASE "-O3")
+
+    add_compile_definitions(
+            [[JUCE_ANDROID=1]]
+    )
+
+endif (ANDROID)
+
+add_compile_definitions(
+        [[JUCEAAP_USE_UNIQUE_ID=1]]
+        [[JUCEAAP_HOSTED_PARAMETER=1]]
+        [[JUCE_DONT_AUTO_OPEN_MIDI_DEVICES_ON_MOBILE=1]]
+        [[JUCEAAP_HAVE_AUDIO_PLAYHEAD_NEW_POSITION_INFO=1]]
+)
+
+target_link_libraries(Odin2
+        PRIVATE
+        aap_audio_processors
+        ${log}
+        ${android}
+        ${glesv2}
+        ${egl}
+        ${cpufeatures_lib}
+)
+# <-- end JUCEAAP specific
diff --git a/Source/PluginEditor.cpp b/Source/PluginEditor.cpp
index e02314c8..fc7afa89 100644
--- a/Source/PluginEditor.cpp
+++ b/Source/PluginEditor.cpp
@@ -128,13 +128,13 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_patch_browser.forceValueTreeLambda = [&]() { forceValueTreeOntoComponents(true); };
 
 	m_reset.onClick = [&]() {
-		if (AlertWindow::showOkCancelBox(AlertWindow::WarningIcon,
+		if (true/*AlertWindow::showOkCancelBox(AlertWindow::WarningIcon,
 		                                 "Reset Synth",
 		                                 "This will reset the synth to its initial state and you will lose "
 		                                 "your patch!",
 		                                 {},
 		                                 {},
-		                                 {})) {
+		                                 {})*/) {
 
 			// replace stream with patch from binary data
 			MemoryInputStream init_stream(BinaryData::init_patch_odin, BinaryData::init_patch_odinSize, false);
@@ -200,7 +200,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_osc1_dropdown.onClick = [&]() {
 		m_osc1_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_osc1_dropdown.setToggleState(false, dontSendNotification);
-		setOsc1Plate(m_osc_dropdown_menu.show());
+		m_osc_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) { setOsc1Plate(v); });
 	};
 	addAndMakeVisible(m_title_button);
 
@@ -220,7 +220,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_osc2_dropdown.onClick = [&]() {
 		m_osc2_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_osc2_dropdown.setToggleState(false, dontSendNotification);
-		setOsc2Plate(m_osc_dropdown_menu.show());
+		m_osc_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) { setOsc2Plate(v); });
 	};
 	m_osc2_dropdown.setTooltip("Select the module to\nbe used for oscillator 2");
 	addAndMakeVisible(m_osc2_dropdown);
@@ -233,7 +233,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_osc3_dropdown.onClick = [&]() {
 		m_osc3_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_osc3_dropdown.setToggleState(false, dontSendNotification);
-		setOsc3Plate(m_osc_dropdown_menu.show());
+		m_osc_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) {setOsc3Plate(v);});
 	};
 	m_osc3_dropdown.setTooltip("Select the module to\nbe used for oscillator 3");
 	addAndMakeVisible(m_osc3_dropdown);
@@ -246,7 +246,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_filter1_dropdown.onClick = [&]() {
 		m_filter1_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_filter1_dropdown.setToggleState(false, dontSendNotification);
-		setFilter1Plate(m_filter_dropdown_menu.show());
+		m_filter_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) { setFilter1Plate(v); });
 	};
 	m_filter1_dropdown.setTooltip("Select the module to\nbe used for filter 1");
 	addAndMakeVisible(m_filter1_dropdown);
@@ -259,7 +259,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_filter2_dropdown.onClick = [&]() {
 		m_filter2_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_filter2_dropdown.setToggleState(false, dontSendNotification);
-		setFilter2Plate(m_filter_dropdown_menu.show());
+		m_filter_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) {setFilter2Plate(v);});
 	};
 	m_filter2_dropdown.setTooltip("Select the module to\nbe used for filter 2");
 	addAndMakeVisible(m_filter2_dropdown);
@@ -272,7 +272,7 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 	m_filter3_dropdown.onClick = [&]() {
 		m_filter3_dropdown.setState(Button::ButtonState::buttonNormal);
 		m_filter3_dropdown.setToggleState(false, dontSendNotification);
-		setFilter3Plate(m_filter_dropdown_menu.show());
+		m_filter_dropdown_menu.showMenuAsync(PopupMenu::Options{}, [&](int v) { setFilter3Plate(v); } );
 	};
 	m_filter3_dropdown.setTooltip("Select the module to\nbe used for filter 3");
 	addAndMakeVisible(m_filter3_dropdown);
@@ -366,7 +366,8 @@ OdinEditor::OdinEditor(OdinAudioProcessor &p_processor, AudioProcessorValueTreeS
 
 		OdinMenuFeels laf;
 		menu.setLookAndFeel(&laf);
-		const auto ret = menu.show();
+        auto ret = 0;
+		menu.showMenuAsync({}, [&](int v) { ret = v; });
 		menu.setLookAndFeel(nullptr);
 
 		if (ret == 0)
diff --git a/Source/PluginProcessorMisc.cpp b/Source/PluginProcessorMisc.cpp
index 13e4a2bb..a78fa3f6 100644
--- a/Source/PluginProcessorMisc.cpp
+++ b/Source/PluginProcessorMisc.cpp
@@ -149,7 +149,7 @@ void OdinAudioProcessor::setStateInformation(const void *data, int sizeInBytes)
 			int patch_migration_version_read =
 			    xmlState->getChildByName("misc")->getIntAttribute("patch_migration_version");
 			if (patch_migration_version_read > ODIN_PATCH_MIGRATION_VERSION) {
-				AlertWindow::showMessageBox(
+				AlertWindow::showMessageBoxAsync(
 				    AlertWindow::AlertIconType::WarningIcon,
 				    "You are trying to load a project which was saved with a newer Version of Odin2. Please go to "
 				    "www.thewavewarden.com and download the newest version to properly use this project!",
diff --git a/Source/audio/Oscillators/Wavetables/Tables/WavetableData.cpp b/Source/audio/Oscillators/Wavetables/Tables/WavetableData.cpp
index 7f8d3c47..20ae47bf 100644
--- a/Source/audio/Oscillators/Wavetables/Tables/WavetableData.cpp
+++ b/Source/audio/Oscillators/Wavetables/Tables/WavetableData.cpp
@@ -15,6 +15,9 @@
 
 #include "WavetableData.h"
 
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wimplicit-float-conversion"
+
 const float wavetable_data[NUMBER_OF_WAVETABLES][SUBTABLES_PER_WAVETABLE][WAVETABLE_LENGTH] = {
     {{0,          0.0122715,  0.0245412,  0.0368072,  0.0490677,   0.0613207,  0.0735646,  0.0857973,  0.0980171,
       0.110222,   0.122411,   0.134581,   0.14673,    0.158858,    0.170962,   0.18304,    0.19509,    0.207111,
@@ -409996,4 +409999,6 @@ const float (*getWavetableData())[NUMBER_OF_WAVETABLES][SUBTABLES_PER_WAVETABLE]
 }
 const float *getOneSubTable(int p_wavetable, int p_subtable) {
 	return wavetable_data[p_wavetable][p_subtable];
-}
\ No newline at end of file
+}
+
+#pragma clang diagnostic pop
diff --git a/Source/gui/BrowserEntry.cpp b/Source/gui/BrowserEntry.cpp
index a7ebe135..3a8c00c6 100644
--- a/Source/gui/BrowserEntry.cpp
+++ b/Source/gui/BrowserEntry.cpp
@@ -26,7 +26,7 @@ BrowserEntry::BrowserEntry(String p_text) : m_text(p_text) {
 
 	m_rename_editor.onReturnKey = [&]() {
 		if (m_rename_editor.getText().isEmpty()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon, "Empty Name", "Please input a valid name!", "Ok");
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Empty Name", "Please input a valid name!", "Ok");
 			return;
 		}
 		hideRenameEditor();
diff --git a/Source/gui/JsonGuiProvider.cpp b/Source/gui/JsonGuiProvider.cpp
index 108914b7..4c11862d 100644
--- a/Source/gui/JsonGuiProvider.cpp
+++ b/Source/gui/JsonGuiProvider.cpp
@@ -34,9 +34,13 @@ void JsonGuiProvider::saveToFile() {
 
 juce::Rectangle<int> JsonGuiProvider::getBounds(const std::string &p_name, bool p_for_knob) {
 	if (!m_data.contains(p_name)) {
+		/*
 		m_alert_window.setMessage(p_name);
 		m_alert_window.runModalLoop();
 		setBoundsUnscaled(p_name, m_gui_editor.getData());
+		 */
+		std::cerr << "Odin2 error (removing runModalLoop): " << p_name << std::endl;
+		return getBoundsUnscaled(p_name);
 	}
 
 	auto zoom_size = ConfigFileManager::getInstance().getOptionGuiScale();
diff --git a/Source/gui/NumberSelectorWithText.cpp b/Source/gui/NumberSelectorWithText.cpp
index 1eb9d912..2741920f 100644
--- a/Source/gui/NumberSelectorWithText.cpp
+++ b/Source/gui/NumberSelectorWithText.cpp
@@ -33,11 +33,15 @@ NumberSelectorWithText::NumberSelectorWithText(NumberSelector::Type p_type) : Nu
 			// b) make ordering inside menu the same as up/down buttons
 			m_dropdown.addItem(9999 - m_legal_values[index], valueToText(m_legal_values[index]));
 		}
-		int selected = 9999 - m_dropdown.show();
-		if (selected != 9999) {
-			DBG(selected);
-			setValue(selected);
-		}
+		PopupMenu::Options options{};
+		std::function<void(int)> cb = [&](int v) {
+			int selected = 9999 - v;
+			if (selected != 9999) {
+				DBG(selected);
+				setValue(selected);
+			}
+		};
+		m_dropdown.showMenuAsync(options, cb);
 	};
 
 	addAndMakeVisible(m_display);
diff --git a/Source/gui/PatchBrowser.cpp b/Source/gui/PatchBrowser.cpp
index 0fb667f5..432bca9a 100644
--- a/Source/gui/PatchBrowser.cpp
+++ b/Source/gui/PatchBrowser.cpp
@@ -144,7 +144,8 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 		String absolute_path = m_soundbank_selector.getDirectory() + File::getSeparatorString() + p_string;
 
 		if (AlertWindow::showOkCancelBox(
-		        AlertWindow::WarningIcon, "Delete Category", "Are you sure you want to delete the Soundbank " + p_string + " and ALL the presets in it?", {}, {}, {})) {
+		        AlertWindow::WarningIcon, "Delete Category", "Are you sure you want to delete the Soundbank " + p_string + " and ALL the presets in it?", {}, {}, {}, {})) {
+
 			File file_to_delete(absolute_path);
 			if (file_to_delete.deleteRecursively()) {
 				DBG("Directory was deleted!");
@@ -169,7 +170,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 		String absolute_path = m_category_selector.getDirectory() + File::getSeparatorString() + p_string;
 
 		if (AlertWindow::showOkCancelBox(
-		        AlertWindow::WarningIcon, "Delete Category", "Are you sure you want to delete the category " + p_string + " and ALL the presets in it?", {}, {}, {})) {
+		        AlertWindow::WarningIcon, "Delete Category", "Are you sure you want to delete the category " + p_string + " and ALL the presets in it?", {}, {}, {}, {})) {
 			File file_to_delete(absolute_path);
 			if (file_to_delete.deleteRecursively()) {
 				DBG("DBG: Directory was deleted!");
@@ -190,7 +191,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 		DBG("DELETE patch: " + m_patch_selector.getDirectory() + File::getSeparatorString() + p_string);
 		String absolute_path = m_patch_selector.getDirectory() + File::getSeparatorString() + p_string;
 
-		if (AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "Delete Preset", "Are you sure you want to delete preset " + p_string + "?", {}, {}, {})) {
+		if (AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "Delete Preset", "Are you sure you want to delete preset " + p_string + "?", {}, {}, {}, {})) {
 			File file_to_delete(absolute_path);
 			if (file_to_delete.deleteFile()) {
 				DBG("file was deleted!");
@@ -212,7 +213,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		//check whether file already exists
 		if (file_to_write.existsAsFile()) {
-			if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "Preset already exists!", "Are you sure you want to overwrite it?", {}, {}, {}))) {
+			if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "Preset already exists!", "Are you sure you want to overwrite it?", {}, {}, {}, {}))) {
 				//user selected cancel
 				return;
 			}
@@ -238,7 +239,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		//check whether directory already exists
 		if (dir_to_create.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon, "Category Exists!", "The category you're trying to create already exists!", "Ok");
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Category Exists!", "The category you're trying to create already exists!", "Ok");
 			return;
 		}
 		if (dir_to_create.createDirectory()) {
@@ -256,7 +257,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		//check whether directory already exists
 		if (dir_to_create.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon, "Soundbank Exists!", "The soundbank you're trying to create already exists!", "Ok");
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "Soundbank Exists!", "The soundbank you're trying to create already exists!", "Ok");
 			return;
 		}
 		if (dir_to_create.createDirectory()) {
@@ -299,7 +300,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 				//check whether file already exists
 				if (file_to_write.existsAsFile()) {
-					if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "File already exists!", "Are you sure you want to overwrite it?", {}, {}, {}))) {
+					if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "File already exists!", "Are you sure you want to overwrite it?", {}, {}, {}, {}))) {
 						//user selected cancel
 						return;
 					}
@@ -318,7 +319,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 	m_soundbank_selector.onExport = [&](String p_directory) {
 		File soundbank_file(m_category_selector.getDirectory());
 		if (!soundbank_file.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon, "No soundbank selected!", "You appear to not have selected any soundbank.", "Bummer");
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "No soundbank selected!", "You appear to not have selected any soundbank.", "Bummer");
 		}
 
 		auto suggested_dir = ConfigFileManager::getInstance().getOptionSoundbankDir();
@@ -343,7 +344,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 				//check whether file already exists
 				if (file_to_write.existsAsFile()) {
-					if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "File already exists!", "Are you sure you want to overwrite it?", {}, {}, {}))) {
+					if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon, "File already exists!", "Are you sure you want to overwrite it?", {}, {}, {}, {}))) {
 						//user selected cancel
 						return;
 					}
@@ -389,7 +390,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		File move_target(p_directory + File::getSeparatorString() + p_new_name + ".odin");
 		if (move_target.existsAsFile()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Preset already exists!",
 			                            "The preset you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -417,7 +418,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		File move_target(p_directory + File::getSeparatorString() + p_new_name);
 		if (move_target.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Category already exists!",
 			                            "The category you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -450,7 +451,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		File move_target(p_directory + File::getSeparatorString() + p_new_name);
 		if (move_target.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Soundbank already exists!",
 			                            "The soundbank you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -483,7 +484,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		File move_target(target_absolute);
 		if (move_target.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Category already exists!",
 			                            "The category you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -512,7 +513,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 		File move_target(m_category_selector.getDirectory() + File::getSeparatorString() + p_target_cat + File::getSeparatorString() + p_file + ".odin");
 
 		if (move_target.existsAsFile()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Preset already exists!",
 			                            "The preset you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -538,7 +539,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 
 		File move_target(target_absolute);
 		if (move_target.isDirectory()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Category already exists!",
 			                            "The category you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -566,7 +567,7 @@ PatchBrowser::PatchBrowser(OdinAudioProcessor &p_processor, AudioProcessorValueT
 		File copy_target(m_category_selector.getDirectory() + File::getSeparatorString() + p_target_cat + File::getSeparatorString() + p_file + ".odin");
 
 		if (copy_target.existsAsFile()) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Preset already exists!",
 			                            "The preset you're trying to create already exists, please choose another name "
 			                            "or remove the other one.",
@@ -610,7 +611,7 @@ void PatchBrowser::loadPatchFromOpenedFileStream(juce::FileInputStream &p_file_s
 	std::string version_string;
 	if (checkForBiggerVersion(p_file_stream, version_string)) {
 		//abort with icon
-		AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+		AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 		                            "Cannot load patch!",
 		                            "The bad news: You cannot load this patch, because you are on version " + ODIN_VERSION_STRING +
 		                                ".\nThe good news: The patch you're trying to load was created on version " + version_string +
@@ -903,7 +904,7 @@ void PatchBrowser::loadPatchWithFileBrowserAndCopyToCategory(String p_directory)
 				ConfigFileManager::getInstance().saveDataToFile();
 
 				if (copy_target.existsAsFile()) {
-					AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+					AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 					                            "Preset already exists!",
 					                            "The preset you're trying to import already exists in this category, "
 					                            "please choose a different category name "
diff --git a/Source/gui/PatchBrowserSelector.cpp b/Source/gui/PatchBrowserSelector.cpp
index fef6b8a3..314589be 100644
--- a/Source/gui/PatchBrowserSelector.cpp
+++ b/Source/gui/PatchBrowserSelector.cpp
@@ -139,7 +139,7 @@ void PatchBrowserSelector::paint(Graphics &g) {
 void PatchBrowserSelector::applyInputField() {
 
 	if (m_input_field.getText().isEmpty()) {
-		AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon, "No name provided!", "Please enter a name into the text field!", "Jeeez, okay...");
+		AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon, "No name provided!", "Please enter a name into the text field!", "Jeeez, okay...");
 	} else {
 		onCreateNewFile(getDirectory() + File::getSeparatorString() + m_input_field.getText());
 		m_input_field_active = false;
@@ -397,8 +397,7 @@ void PatchBrowserSelector::generateContent() {
 					unhighlightAllEntries();
 				};
 				m_entries.back()->onRightClick = [&, return_string, file_index]() {
-					int selected = m_menu.show();
-
+					m_menu.showMenuAsync(PopupMenu::Options{}, [&](int selected) {
 					DBG_VAR(file_index);
 
 					if (selected == PATCH_BROWSER_MENU_ENTRY_DELETE) {
@@ -413,6 +412,7 @@ void PatchBrowserSelector::generateContent() {
 					} else if (selected >= PATCH_BROWSER_MENU_COPY_OFFSET) {
 						onCopy(m_entries[file_index]->getText(), getCopyFileString(selected));
 					}
+					});
 				};
 				m_entries.back()->applyRenaming = [&](String p_old_name, String p_new_name) { applyRenamingSelector(getDirectory(), p_old_name, p_new_name); };
 			}
diff --git a/Source/gui/SyncTimeSelector.cpp b/Source/gui/SyncTimeSelector.cpp
index 3e9d73e2..684f8640 100644
--- a/Source/gui/SyncTimeSelector.cpp
+++ b/Source/gui/SyncTimeSelector.cpp
@@ -40,7 +40,7 @@ SyncTimeSelector::SyncTimeSelector(UIAssets::Indices p_asset) :
 	m_dropdown.addItem(120, "1 / 32");
 	m_dropdown.addItem(130, "1 / 32T");
 
-	m_display.onMouseDown = [&]() { setValuesFromDropdown(m_dropdown.show()); };
+	m_display.onMouseDown = [&]() { m_dropdown.showMenuAsync({}, [&](auto code) { setValuesFromDropdown(code); }); };
 
 	m_dropdown.setLookAndFeel(&m_menu_feels);
 
diff --git a/Source/gui/TuningComponent.cpp b/Source/gui/TuningComponent.cpp
index 5bf443d6..93cb4d50 100644
--- a/Source/gui/TuningComponent.cpp
+++ b/Source/gui/TuningComponent.cpp
@@ -114,6 +114,7 @@ void TuningComponent::exportSCLFileWithFileBrowser() {
 
 			    //check whether file already exists
 			    if (file_to_write.existsAsFile()) {
+					/*
 				    if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon,
 				                                       "File already exists!",
 				                                       "Are you sure you want to overwrite it?",
@@ -123,6 +124,7 @@ void TuningComponent::exportSCLFileWithFileBrowser() {
 					    //user selected cancel
 					    return;
 				    }
+				    */
 			    }
 			    FileOutputStream file_stream(file_to_write);
 
@@ -137,7 +139,7 @@ void TuningComponent::exportSCLFileWithFileBrowser() {
 				    file_stream.flush();
 				    //savePatchInOpenedFileStream(file_stream);
 			    } else {
-				    AlertWindow::showMessageBox(
+				    AlertWindow::showMessageBoxAsync(
 				        AlertWindow::AlertIconType::WarningIcon,
 				        "Unable to save file!",
 				        "Please make sure you have proper write-permissions for the selected directory!");
@@ -174,6 +176,7 @@ void TuningComponent::exportKBMFileWithFileBrowser() {
 
 			    //check whether file already exists
 			    if (file_to_write.existsAsFile()) {
+					/*
 				    if (!(AlertWindow::showOkCancelBox(AlertWindow::WarningIcon,
 				                                       "File already exists!",
 				                                       "Are you sure you want to overwrite it?",
@@ -183,6 +186,7 @@ void TuningComponent::exportKBMFileWithFileBrowser() {
 					    //user selected cancel
 					    return;
 				    }
+				    */
 			    }
 			    FileOutputStream file_stream(file_to_write);
 
@@ -196,7 +200,7 @@ void TuningComponent::exportKBMFileWithFileBrowser() {
 				    file_stream << String(m_processor.m_tuning.keyboardMapping.rawText);
 				    file_stream.flush();
 			    } else {
-				    AlertWindow::showMessageBox(
+				    AlertWindow::showMessageBoxAsync(
 				        AlertWindow::AlertIconType::WarningIcon,
 				        "Unable to save file!",
 				        "Please make sure you have proper write-permissions for the selected directory!");
@@ -254,7 +258,7 @@ void TuningComponent::importKBMFromFileBrowser(String p_directory,
 		try {
 			kbm_from_file = Tunings::readKBMFile(file_name.toStdString());
 		} catch (...) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Unable to read keyboard mapping!",
 			                            "Please make sure the file is a proper keyboard mapping file!");
 			return;
@@ -297,7 +301,7 @@ void TuningComponent::importSCLFromFileBrowser(String p_directory,
 		try {
 			scale_from_file = Tunings::readSCLFile(file_name.toStdString());
 		} catch (...) {
-			AlertWindow::showMessageBox(AlertWindow::AlertIconType::WarningIcon,
+			AlertWindow::showMessageBoxAsync(AlertWindow::AlertIconType::WarningIcon,
 			                            "Unable to read Scala file!",
 			                            "Please make sure the file is a proper Scala file!");
 			return;
diff --git a/Source/gui/WaveformSelectorComponent.cpp b/Source/gui/WaveformSelectorComponent.cpp
index 858f10af..6d48b5f1 100644
--- a/Source/gui/WaveformSelectorComponent.cpp
+++ b/Source/gui/WaveformSelectorComponent.cpp
@@ -36,7 +36,7 @@ WaveformSelectorComponent::WaveformSelectorComponent(bool p_buttons_right) : m_u
 	m_down.setColour(juce::DrawableButton::ColourIds::backgroundOnColourId, juce::Colour());
 	m_down.onClick = [&]() { increment(); };
 
-	m_display.onMouseDown = [&]() { setValue(m_menu.show()); };
+	m_display.onMouseDown = [&]() { m_menu.showMenuAsync({}, [&](auto code) { setValue(code); }); };
 
 	m_buttons_right = p_buttons_right;
 
diff --git a/libs/JUCE b/libs/JUCE
--- a/libs/JUCE
+++ b/libs/JUCE
@@ -1 +1 @@
-Subproject commit 2f980209cc4091a4490bb1bafc5d530f16834e58
+Subproject commit 2f980209cc4091a4490bb1bafc5d530f16834e58-dirty
